internal/util/update/update_test.go

```diff
@@ -789,6 +790,210 @@ func TestCommand_Run_badStrategy(t *testing.T) {
 	}
 }
 
+func TestCommand_Run_subpackages(t *testing.T) {
+	testCases := []struct {
+		name            string
+		initialUpstream *dataset.Pkg
+		updatedUpstream *dataset.Pkg
+		updatedLocal    *dataset.Pkg
+		expectedLocal   *dataset.Pkg
+	}{
+		{
+			// TODO(mortent): This does not handle Kptfiles correctly.
+			name: "update fetches any new subpackages",
+			initialUpstream: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+				),
+			updatedUpstream: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile().
+						WithSubPackages(
+							dataset.NewPackage("nestedbar").
+								WithKptfile(),
+						),
+					dataset.NewPackage("zork").
+						WithKptfile(),
+				),
+			expectedLocal: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile().
+						WithSubPackages(
+							dataset.NewPackage("nestedbar"),
+						),
+					dataset.NewPackage("zork"),
+				),
+		},
+		{
+			name: "local updates remain after noop update",
+			initialUpstream: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+				),
+			updatedLocal: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+					dataset.NewPackage("zork").
+						WithKptfile(),
+				),
+			expectedLocal: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+					dataset.NewPackage("zork").
+						WithKptfile(),
+				),
+		},
+		{
+			name: "non-overlapping additions in both upstream and local is ok",
+			initialUpstream: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+				),
+			updatedUpstream: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+					dataset.NewPackage("zork").
+						WithKptfile(),
+				),
+			updatedLocal: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+					dataset.NewPackage("abc").
+						WithKptfile(),
+				),
+			expectedLocal: dataset.NewPackage("foo").
+				WithKptfile().
+				WithSubPackages(
+					dataset.NewPackage("bar").
+						WithKptfile(),
+					dataset.NewPackage("zork"),
+					dataset.NewPackage("abc").
+						WithKptfile(),
+				),
+		},
+		{
+			// TODO(mortent): This probably shouldn't work.
```

Comment:
I think we should see this merge more at file level and not package level to be simple and accurate. Treating a package as a unit is not what we want to do here. If upstream `bar` is different from local `bar`, the files in both `bar` packages must be merged. WDYT?

Reply:
I don't think so. If a package of name `bar` is added independently in upstream and in local, it seems weird to just merge them. In a case where they are completely different, the result of a merge will be all the resources from both combined. That doesn't seem correct.